EIPS/eip-3005.md

```diff
@@ -0,0 +1,340 @@
+---
+eip: 3005
+title: Batched meta transactions
+author: Matt (@defifuture)
+discussions-to: https://ethereum-magicians.org/t/eip-3005-the-economic-viability-of-batched-meta-transactions/4673 
+status: Draft
+type: Standards Track
+category: ERC
+created: 2020-09-25
+---
+
+## Simple Summary
+
+Defines an extension function for ERC-20 (and other fungible token standards), which allows receiving and processing a batch of meta transactions.
+
+![](../assets/eip-3005/meta-txs-directly-to-token-smart-contract.png)
+
+## Abstract
+
+This EIP proposes a new function called `processMetaBatch()` that extends any fungible token standard, and enables batched meta transactions coming from many senders in one on-chain transaction. 
+
+The function must be able to receive multiple meta transactions data and process it. This means validating the data and the signature, before proceeding with token transfers based on the data.
+
+The function enables senders to make gasless transactions, while reducing the relayer's gas cost due to batching.
+
+## Motivation
+
+Meta transactions have proven useful as a solution for Ethereum accounts that don't have any ether, but hold ERC-20 tokens and would like to transfer them (gasless transactions).
+
+The current meta transaction relayer implementations only allow relaying one meta transaction at a time. Some also allow batched meta transactions from the same sender. But none offers batched meta transactions from **multiple** senders.
+
+The motivation behind this EIP is to find a way to allow relaying batched meta transactions from **many senders** in **one on-chain transaction**, which also **reduces the total gas cost** that a relayer needs to cover.
+
+## Specification
+
+### Meta transaction data
+
+In order to successfully validate and transfer tokens, the proposed function needs to process the following data about a meta transaction:
+
+- sender address
+- receiver address
+- token amount
+- relayer fee
+- a (meta tx) nonce
+- a timestamp or a block number (which represents a due date to process a meta tx)
+- a token address
+- a relayer address
+- a signature
```

Comment:
In order for this standard to be implementable, this will need to be much more strictly defined.  Often people will include a Solidity interface definition for the new function, though you also could strictly define the full calldata layout if you preferred.

Reply:
Do you think I should write this the way it's defined in EIP-2612 (the solidity pseudocode in the Specification)? Or are there any better EIP examples to learn from (applicable to my case, of course)? 🙂

Reply:
I'm looking now at EIP-20 and EIP-777 - do these two have a good Specification section?

Reply:
EIP-20 and EIP-777 would be a reasonable place to start.  The goal is to make it so someone can read the Specification section *only* and get exactly enough information required to implement something that is compatible with the standard.  "Here are the things you must do to be considered compatible."